Refactor tkey-verification: new ways of accessing compiled-in assets

mchack-work opened the pull request:
Refactoring to:

- Include the device binaries in the repo directly. Remove the build scripts for verisigner. Use the ordinary tkey-device-signer for new stuff. If a user wants to build verisigner, they have to use old tags and the instructions there.

- Use ordinary tkeyclient package instead of internal version.

- Use the ordinary tkeysign package to communicate with both signer, these days updated to be able to get a firmware hash, and the old verisigner.

- Use new way of accessing internal assets like: vendor public keys, known firmwares, and the app binaries. We're now more idiomatic in the access and use something like `foos := NewFoo()` with methods on `foo` to do something with them.

- Support multiple vendor keys at the same time for verification.

- Make the external command `show-pubkey` into just another command in `tkey-verification`.

- Error handling much changed. Still a little different in verify.go compared to the rest of the program. Not sure what we do about that. Perhaps just leave it.

A good way to start to understand the refactoring is to look in the new [Implementation notes](https://github.com/tillitis/tkey-verification/blob/signer/doc/implementation-notes.md)

Closes #18

dehanj commented on cmd/tkey-verification/verify.go:
```diff
@@ -15,114 +15,129 @@ import (
+		os.Exit(code)
+	}
+
+	fmt.Printf("tk: %#v\n", tk)
```
Is this a leftover debug print?

mchack-work replied:
Yes.

mchack-work commented:
- Updated the description of what this PR does.
- Rebased on main.
- Squashed and rewored many commits.

Please have a look @dehanj. Do you think we should squash the entire thing into something like the new PR description or do we leave the individual commits? Some of them are mostly small fixes.

Not sure what to do about the first commit... It's from the tkey-libs branch which has been abondened. Do we even try to remove it?

dehanj reviewed (commented):
First set of comments

dehanj commented:
When using `tkey-verification verify` -d /path/to/dir` with a local copy of the verification data, if I change the name of the apptag (to for example "verisigner-v0.0.4"), nothing happens. From what I can see it never uses the apptag for anything. Should it?

Is apptag used for anything today, or is it simply a human readable tag?

dehanj commented on cmd/tkey-verification/main.go:
```diff
@@ -81,35 +57,25 @@ Known firmwares:
-
-		le.Printf("%s\n\nFlags:\n%s\n%s", desc, pflag.CommandLine.FlagUsagesWrapped(86), builtWith)
-	}
+	pflag.Usage = usage
```
There is no `--speed` flag. Right now this prevents tkey-verification to be runnable on macos, since QEMU for some reason (a bug I guess) uses baudrate of 9600 baud.

mchack-work commented on cmd/tkey-verification/appbins.go:
```diff
@@ -0,0 +1,136 @@
+// NewAppBins initializes the embedded device apps. It expects the
+// hash digest of the latest device app as latestHash. This is used
+// for device signature during provisioning.
+func NewAppBins(latestHash string) (AppBins, error) {
```
This is indeed confusing. I've changed it to let it be configurable in config files instead. I also changed the vendor signing app to be configurable.

The apps need to be built in, but the values of what is used doesn't.

Please look at the new configuration file examples and the new structs in config.go.

mchack-work commented on cmd/tkey-verification/vendorpubkeys.go:
```diff
@@ -0,0 +1,118 @@
+// Copyright (C) Tillitis AB
```
Perhaps this file is the only one that's correct?

https://daniel.haxx.se/blog/2023/01/08/copyright-without-years/

dehanj replied:
Maybe. Would make it a lot easier.

mchack-work replied:
Probably. We will probably need to update the years in many places. This PR touches many files, not all of which have 2024 as copyright year.

dehanj replied:
I'll fix this by updating all applicable files in a commit. Then we can discuss further if we are to drop the year on these notices in the files, but keep it for the one in the readme.

mchack-work commented:
> When using `tkey-verification verify -d /path/to/dir` with a local copy of the verification data, if I change the name of the apptag (to for example "verisigner-v0.0.4"), nothing happens. From what I can see it never uses the apptag for anything. Should it?
> 
> Is apptag used for anything today, or is it simply a human readable tag?

It's only a human readable tag now. Should be documented as such, so good point.

dehanj commented on README.md:
```diff
@@ -78,183 +93,166 @@ signer signature during provisioning *must* be used when verifying it.
 If a different signer is used then the device public key will not
 match even if the TKey is the same. A verifier must check the "tag"
```
Fixed.

dehanj commented on README.md:
```diff
@@ -2,34 +2,47 @@
+provisioned, typically by [Tillitis](https://tillitis.se/) the
+original vendor.
+
+*Note well*: If your TKey has been provisioned by someone else, like
```
fixed.

dehanj commented on internal/tkey/tkey.go:
```diff
@@ -1,302 +1,170 @@
+func NewTKey(devPath string, verbose bool) (*TKey, error) {
 	if !verbose {
-		tk1.SilenceLogging()
+		tkeyclient.SilenceLogging()
```
We are not fixing this now.

dehanj commented on README.md:
````diff
@@ -78,183 +93,166 @@ signer signature during provisioning *must* be used when verifying it.
+$ cat other-pubkey.txt >> cmd/tkey-verification/vendor-signing-pubkeys.txt
 $ make
 ```
 
````
Fixed, renamed to "Verification file". 
This term i already used in different places. 
Also considered "Verification data".

dehanj commented on cmd/tkey-verification/vendorpubkeys.go:
```diff
@@ -0,0 +1,118 @@
+		}
+
+		if len(fields) != 3 {
+			return vendorKeys, SimpleParseError{msg: "Expected 3 space-separated fields: pubkey in hex, verisigner-app tag, and its hash in hex"}
```
fixed.

dehanj commented on cmd/tkey-verification/vendorpubkeys.go:
```diff
@@ -0,0 +1,118 @@
+		if _, ok := appBins.Bins[appHashHex]; ok {
+			appBin = appBins.Bins[appHashHex]
+			if appBin.Tag != tag {
+				return vendorKeys, EqualError{"embedded app tag", "public key tag"}
```
Fixed.

dehanj commented:
Fixed most of my comments. Maybe some more squashing can be done, but waiting with that until the speed is included.

Otherwise I start to think that it looks good. The functionality has been verified.

dehanj reviewed (approved):
Managed to squash another 6 commits.

At this stage it looks good to me.
I have verified:
- multiple vendor signing key support, and hence do signing and verifying
- that it does not break verification files out in the field
- Setting up the a test environment with QEMU
- the new speed command
- Added the correct year for the copyright notices
- Revised part of readme (squashed)

Ready to merge.
